### ISSUE TITLE
BGP peers module for Junos devices

### ISSUE THREAD

**Issue by mog01** [librenms/librenms#5715]
DO NOT DELETE THIS INFORMATION.

> Please read this information carefully.

GitHub issues is for feature requests or bugs, please do not post issues asking for help or how to do X, Y or Z. 
You can use our irc channel ##librenms on freenode to ask questions or our [community site](https://community.librenms.org).

Please confirm each of the sections below by putting an `x` in the box like `[x]`.

- [x] Is your install up to date? [Updating your install](http://docs.librenms.org/General/Updating/)
      Please do not submit an issue if your install is not up to date within the last 24 hours or on a stable monthly release.
- [x] Please include all of the information between the `====================================` section of `./validate.php` which you can run from the cli.
- [ ] Unless your issue is for a WebUI fix or feature then please provide ALL info asked for [here](http://docs.librenms.org/Support/FAQ/#faq20).
- [x] Please provide as much detail as possible.
- [x] Please do NOT post more than 10 lines of debug information here, use a pastebin service or GitHub Gists.

Component | Version
--------- | -------
LibreNMS  | 8aa2480c05e20f2bd0b975925cef34e5d0a8038f
DB Schema | 157
PHP       | 7.0.12
MySQL     | 5.5.47-MariaDB
RRDTool   | 1.4.8
SNMP      | NET-SNMP 5.7.2

When polling Junos devices for BGP peering information, librenms returns with no object found. Please find output from discovery.php and poller.php. Unfortunately the snmpbulkwalk output is too large for pastebin.com to accept :(

discovery: http://pastebin.com/f3MwLb7j
poller: http://pastebin.com/GGQWFnk7

**Comment by laf** [comment]
You will have to find a way to get the walk output to us

**Comment by mog01** [comment]
Hi, presumably I'm not the first person to encounter this problem. How have people solved it in the past please? Or what options do you have fore receiving files? Maybe an email address?

**Comment by mog01** [comment]
Here's another walk specifying 1.3.6.1.4.1.2636 as the prefix:

[junos-subwalk.txt](https://github.com/librenms/librenms/files/821612/junos-subwalk.txt)

**Comment by laf** [comment]
Unfortunately this doesn't seem to have the data from the Juniper BGP mib, data should be in .1.3.6.1.4.1.2636.5.1.1.1.5 for the local AS. Can you do a walk and see if you get anything back on that.

**Comment by Tatermen** [comment]
@mog01 Could you state what model Juniper device you have that isn't working? 

For what it's worth, I can confirm that the stock BGP polling works with MX, ACX and SRX devices.

**Comment by mog01** [comment]
Hi, the MX104s are currently running JunOS 13.3R8.7

**Comment by laf** [comment]
No further info provided.

**Comment by mog01** [comment]
Sorry, what? I was asked to state which model of Juniper device we run which is not working as expected. I stated the model and JunOS version in question, as well as a different model and JunOS version in the hope that it might be helpful.

Please explain how the ticket can then be closed (after about 24 days) with a lone comment of "no further info provided"? I believe I have already demonstrated that I am willing to provide as much information as I reasonably can; what further information is requested?

**Comment by laf** [comment]
You've not responded to my last request for more data.

**Comment by Tatermen** [comment]
@mog01 Your snmp walk is entirely missing the Juniper BGP SNMP OIDs. If I compare your snmpwalk to one of my devices, your entire .1.3.6.1.4.1.2636.5 tree is missing, which maps asBGP4-V2-MIB-JUNIPER. For example:

```
.1.3.6.1.4.1.2636.5.1.1.2.1.1.1.1.0.1.172.16.255.0.1.172.16.255.1 = Hex-STRING: AC 10 FF 01
.1.3.6.1.4.1.2636.5.1.1.2.1.1.1.1.0.1.172.16.255.0.1.172.16.255.10 = Hex-STRING: AC 10 FF 0A
.1.3.6.1.4.1.2636.5.1.1.2.1.1.1.1.0.1.172.16.255.0.1.172.16.255.11 = Hex-STRING: AC 10 FF 0B
.1.3.6.1.4.1.2636.5.1.1.2.1.1.1.1.0.1.172.16.255.0.1.172.16.255.13 = Hex-STRING: AC 10 FF 0D
.1.3.6.1.4.1.2636.5.1.1.2.1.1.1.1.0.1.172.16.255.0.1.172.16.255.20 = Hex-STRING: AC 10 FF 14
```

If your device is actually missing this MIB, there's nothing wrong with LibreNMS. Either you've hit a bug in the version of JunOS that you are using and it's not populating those OIDs,  or something in your SNMP configuration is preventing LibreNMS from reading those OIDs.

**Comment by mog01** [comment]
Thanks for your help / replies.
Oh, sorry Laf, didn't see your message mixed in there. Hrm, I think these devices might be using the original SNMP config set up by a predecessor. I will check that and see if that helps.

**Comment by mog01** [comment]
I think I've found out why LibreNMS isn't able to poll our Juniper devices for BGP peering information. I checked Juniper's MIB explorer and found that the BGP info uses a different MIB prefix:

https://apps.juniper.net/mib-explorer/

Name: bgp
OID: 1.3.6.1.2.1.15
Syntax: MODULE-IDENTITY
Description: The MIB module for BGP-4.


If I do an SNMP walk with a prefix of "1.3.6.1.2.1.15" I get lots of information back:

.1.3.6.1.2.1.15.6.1.14.193.xxx.xxx.0.23.xxx.xxx.40.49 = ""
.1.3.6.1.2.1.15.6.1.14.217.xxx.xxx.0.20.xxx.xxx.40.49 = ""
.1.3.6.1.2.1.15.6.1.14.217.xxx.xxx.0.24.xxx.xxx.40.49 = ""

Is it possible to change the BGP module to use a different prefix?

**Comment by laf** [comment]
Data needs to be from 1.3.6.1.2.1.15.3. 1.3.6.1.2.1.15.6 is a table we have no use for.

**Comment by mog01** [comment]
Ah right, I found some stuff from 15.3:

.1.3.6.1.2.1.15.3.1.20.xxx.xxx.254.65 = INTEGER: 30
.1.3.6.1.2.1.15.3.1.20.xxx.xxx.254.69 = INTEGER: 30
.1.3.6.1.2.1.15.3.1.20.xxx.xxx.255.81 = INTEGER: 30

**Comment by laf** [comment]
I'm not sure if that's all you've got but that's just  bgpPeerHoldTimeConfigured which is of no use.

Are you not even seeing ` .1.3.6.1.2.1.15.2` or ` .1.3.6.1.2.1.15.3` but in all honestly, this really does sounds like a question for Juniper. They don't seem to be using either BGP4-MIB or there own for this device.

**Comment by mog01** [comment]
Sorry I'm confused, the prefix I posted was for .1.3.6.1.2.1.15.3; there's lots of entries for that prefix, I just posed a sample few.

The devices in question is an SRX240, I also get data from the same prefix via an MX104 Juniper router.

**Comment by laf** [comment]
A few is of no use, that one you chose is for data we don't use or care about. You need to provide a full walk of all that data otherwise this is issue is just going to go round in circles.

**Comment by mog01** [comment]
I posted a few to demonstrate that the BGP tree is indeed available, although apparently under a different prefix to what LibreNMS expects, based on what Tatermen has experienced.

Juniper's MIB Explorer shows details of all available SNMP information, for example:

https://apps.juniper.net/mib-explorer/search.jsp#object=bgpPeerTable&product=Junos%20OS&release=12.1x46-D65

The few bits of information I pasted simply confirm that the devices / JunOS we're running match up with Juniper's SNMP documentation.

**Comment by Tatermen** [comment]
> although apparently under a different prefix to what LibreNMS expects, based on what Tatermen has experienced.

That's not what I said. I confirmed that LibreNMS is perfectly able to see the BGP tree on the Juniper MX devices that I have at my disposable.  I said something is wrong with your Juniper devices if they're not populating the Juniper BGP MIB. 

From one of our MX104s:

![image](https://user-images.githubusercontent.com/8707177/27152210-0886cfa4-5145-11e7-9707-1ec1bf218100.png)

From a HA pair of SRX240s:

![image](https://user-images.githubusercontent.com/8707177/27152285-4b8c59fe-5145-11e7-97da-e9f5e1d781c3.png)

@laf said earlier that LibreNMS looks in .1.3.6.1.4.1.2636.5.1.1.1.5, which is a Juniper specific BGP MIB - BGP4-V2-MIB-JUNIPER. https://apps.juniper.net/mib-explorer/search.jsp#object=jnxBgpM2Experiment&product=Junos%20OS&release=13.3R10

The MIB you've quoted above is BGP4-MIB, which appears to be a generic MIB for BGP peering information - not the Juniper specific one that LibreNMS is looking for.

tl;dr - I have access to Juniper MX and SRX devices running BGP peerings. LibreNMS can see the BGP peerings. Something is wrong with your Juniper devices - not LibreNMS.

**Comment by mog01** [comment]
What SNMP configuration is being used on your devices please?

I can only assume it's a configuration issue. The BGP information was not previously included in the SNMP view by my predecessor, so I added it - presumably something still must still be missing :(

**Comment by mog01** [comment]
Do I understand correctly that LibreNMS looks for experimental BGP MIB information? Not the generic info?

**Comment by Tatermen** [comment]
We don't use SNMPv3. Ours are configure with bog standard v2.

**Comment by mog01** [comment]
Understood, thanks. I've included oid jnxBgpM2Experiment to the SNMP view, I'll wait a bit to see if anything changes within LibreNMS.

**Comment by laf** [comment]
Wait if you're saying you restrict what data is being sent back then that makes total sense. I don't use juniper but allow:

 bgpLocalAs
jnxBgpM2PeerRemoteAs
jnxBgpM2PeerEntry
jnxBgpM2PrefixCountersTable

You may need more but try with those at least.

**Comment by mog01** [comment]
I added those to one of our SRX240s just now, thanks!
Should LibreNMS pick them up automatically or do I maybe need to initiate a rediscover?

**Comment by laf** [comment]
It will automatically run within around 6 hours, you can kick it off yourself but I guess by now it's either worked or not.
